From: Chuanhong Guo <gch981213@gmail.com>
Date: Thu, 23 Jul 2026 12:00:00 +0800
Subject: [PATCH] net: stmmac: add platform TC hook for SF19A2890 HNAT

The SF19A2890 glue driver owns a hardware NAT block between the DWMAC MAC
and DMA.  Let a platform driver handle TC flowtable requests before falling
back to the generic stmmac classifiers, and advertise NETIF_F_HW_TC when
the platform supplies the hook.

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>
---
--- a/include/linux/stmmac.h
+++ b/include/linux/stmmac.h
@@ -15,6 +15,8 @@
 #include <linux/platform_device.h>
 #include <linux/phylink.h>
 
+enum tc_setup_type;
+
 #define MTL_MAX_RX_QUEUES	8
 #define MTL_MAX_TX_QUEUES	8
 #define STMMAC_CH_MAX		8
@@ -254,6 +256,8 @@
 			       phy_interface_t interface, int speed);
 	void (*fix_mac_speed)(void *priv, int speed, unsigned int mode);
 	int (*fix_soc_reset)(struct stmmac_priv *priv, void __iomem *ioaddr);
+	int (*setup_tc)(struct net_device *ndev, enum tc_setup_type type,
+			void *type_data);
 	int (*serdes_powerup)(struct net_device *ndev, void *priv);
 	void (*serdes_powerdown)(struct net_device *ndev, void *priv);
 	int (*mac_finish)(struct net_device *ndev,
--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
@@ -6179,6 +6179,13 @@
 			   void *type_data)
 {
 	struct stmmac_priv *priv = netdev_priv(ndev);
+	int ret;
+
+	if (priv->plat->setup_tc) {
+		ret = priv->plat->setup_tc(ndev, type, type_data);
+		if (ret != -EOPNOTSUPP)
+			return ret;
+	}
 
 	switch (type) {
 	case TC_QUERY_CAPS:
@@ -7533,9 +7540,8 @@
 			     NETDEV_XDP_ACT_XSK_ZEROCOPY;
 
 	ret = stmmac_tc_init(priv, priv);
-	if (!ret) {
+	if (!ret || priv->plat->setup_tc)
 		ndev->hw_features |= NETIF_F_HW_TC;
-	}
 
 	if ((priv->plat->flags & STMMAC_FLAG_TSO_EN) && (priv->dma_cap.tsoen)) {
 		ndev->hw_features |= NETIF_F_TSO | NETIF_F_TSO6;
--- a/drivers/net/ethernet/stmicro/stmmac/Makefile
+++ b/drivers/net/ethernet/stmicro/stmmac/Makefile
@@ -23,6 +23,7 @@
 obj-$(CONFIG_DWMAC_RENESAS_GBETH) += dwmac-renesas-gbeth.o
 obj-$(CONFIG_DWMAC_ROCKCHIP)	+= dwmac-rk.o
 obj-$(CONFIG_DWMAC_SF19A2890)	+= dwmac-sf19a2890.o
+obj-$(CONFIG_DWMAC_SF19A2890)	+= sf19a2890-hnat.o
 obj-$(CONFIG_DWMAC_RZN1)	+= dwmac-rzn1.o
 obj-$(CONFIG_DWMAC_S32)		+= dwmac-s32.o
 obj-$(CONFIG_DWMAC_SOCFPGA)	+= dwmac-altr-socfpga.o
